# HW03 - Selecting, Training and Evaluating Models

The aim of this exercise is to gain some basic understanding how to select a suitable model for your task based on cross-validation. You will also learn to finetune your model.

Please note that you should read chapter 2 “End-to-End Machine Learning Project” of the course book “Hands-On Machine Learning with Scikit-Learn, Keras, and TensorFlow” 2nd/3rd edition by Aurélien Géron (2019/2023) to complete the tasks of this exercise successfully. Also take a look at the end-to-end machine learning project, it will help you to complete this assignment.

Embed all your theory answers, drawings, source codes and screen shots directly into this document. Always immediately after the relevant question.

Return the document into your return box in ItsLearning environment by the deadline. Please remember to mark the number of points that you should get from this exercise according to your own evaluation. **The maximum number of points available in this exercise is 10**. You can check the task specific points directly from each task.

In addition to the course book, it’s also recommendable to use other book and Internet sources to supplement the information.

Use the Diabetes dataset (same than in Classwork 2):

url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv"

columns = ["Pregnancies", "Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI", "DiabetesPedigreeFunction", "Age", "Outcome"]


## Tasks:

**1) Data Preprocessing** (0.5 points)
- In all tasks: split your code into more than one coding cell (whenever you can or is feasible), so there is e.g. one thing/function in one cell. It helps you in running your code and interpreting possible error/warning messages.
- Import all necessary libraries
- Load the data
- y is "Outcome", rest of the features are in X
- Display X and y (use head())


In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as mp
import seaborn as sns
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV, RandomizedSearchCV

In [4]:
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv"
data = pd.read_csv(url)

In [5]:
data.head()

,6,148,72,35,0,33.6,0.627,50,1
0,1,85,66,29,0,26.6,0.351,31,0
1,8,183,64,0,0,23.3,0.672,32,1
2,1,89,66,23,94,28.1,0.167,21,0
3,0,137,40,35,168,43.1,2.288,33,1
4,5,116,74,0,0,25.6,0.201,30,0


In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 767 entries, 0 to 766
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   6       767 non-null    int64  
 1   148     767 non-null    int64  
 2   72      767 non-null    int64  
 3   35      767 non-null    int64  
 4   0       767 non-null    int64  
 5   33.6    767 non-null    float64
 6   0.627   767 non-null    float64
 7   50      767 non-null    int64  
 8   1       767 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [8]:
columns = [
    "Pregnancies",
    "Glucose",
    "BloodPressure",
    "SkinThickness",
    "Insulin",
    "BMI",
    "DiabetesPedigreeFunction",
    "Age",
    "Outcome"
]

data2 = pd.read_csv(url , names = columns)

In [9]:
data2.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


In [10]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [11]:
data2.describe()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
count,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000
mean,3.845052,120.894531,69.105469,20.536458,79.799479,31.992578,0.471876,33.240885,0.348958
std,3.369578,31.972618,19.355807,15.952218,115.244002,7.884160,0.331329,11.760232,0.476951
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000,0.000000
25%,1.000000,99.000000,62.000000,0.000000,0.000000,27.300000,0.243750,24.000000,0.000000
50%,3.000000,117.000000,72.000000,23.000000,30.500000,32.000000,0.372500,29.000000,0.000000
75%,6.000000,140.250000,80.000000,32.000000,127.250000,36.600000,0.626250,41.000000,1.000000
max,17.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,81.000000,1.000000


In [14]:
data2.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [16]:
(data2 == 0).sum()

Pregnancies                 111
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                     500
dtype: int64

In [17]:
cols_zero = ["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]

data2[cols_zero] = data2[cols_zero].replace(0, np.nan)

In [19]:
(data2 == 0).sum()

Pregnancies                 111
Glucose                       0
BloodPressure                 0
SkinThickness                 0
Insulin                       0
BMI                           0
DiabetesPedigreeFunction      0
Age                           0
Outcome                     500
dtype: int64

In [20]:
data2.isnull().sum()

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

In [22]:
from sklearn.impute import SimpleImputer

imputer = SimpleImputer(strategy = "median")

data2[cols_zero] = imputer.fit_transform(data2[cols_zero])

In [23]:
data2.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [ ]:
y is "Outcome", rest of the features are in X
Display X and y (use head())

In [27]:
X = data2.drop("Outcome", axis = 1)
y = data2["Outcome"]

In [28]:
X.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age
0,6,148.0,72.0,35.0,125.0,33.6,0.627,50
1,1,85.0,66.0,29.0,125.0,26.6,0.351,31
2,8,183.0,64.0,29.0,125.0,23.3,0.672,32
3,1,89.0,66.0,23.0,94.0,28.1,0.167,21
4,0,137.0,40.0,35.0,168.0,43.1,2.288,33


In [29]:
y.head()

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64

**2) Use train_test_split with 80-20 split and random_state=42. Make sure your data is in X_train, X_test, y_train, y_test. Scale the data after the split.** (0.5 points)
- Also find out, why we use capital X but smaller case y in the names (no need to write down the answer here, but just clarify that to yourself).

In [30]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size = 0.2,
    random_state = 42
)

In [31]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)

X_test = scaler.transform(X_test)

**3) Use k-fold cross-validation with linear regression and decision tree (regression) model.**
   - Report RMSE for cross-validation for each. (1 point if done for both, 0.5 points for one)
   - Fit on both models (on the X_train, y_train) and predict (on the X_test), report RMSE for both. (1 point if done for both, 0.5 points for one)


In [32]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error

In [34]:
lin_reg = LinearRegression()
tree_reg = DecisionTreeRegressor(random_state = 42)

In [35]:
lin_scores = cross_val_score(
    lin_reg,
    X_train,
    y_train,
    scoring = "neg_mean_squared_error",
    cv = 5
)

lin_rmse_scores = np.sqrt(-lin_scores)
lin_rmse_scores.mean()

0.3989800903349941

In [36]:
tree_scores = cross_val_score(
    tree_reg,
    X_train,
    y_train,
    scoring = "neg_mean_squared_error",
    cv = 5
)

tree_rmse_scores = np.sqrt(-tree_scores)
tree_rmse_scores.mean()

0.5754013412103671

In [37]:
lin_reg.fit(X_train, y_train)
tree_reg.fit(X_train, y_train)

DecisionTreeRegressor(random_state=42)

In [39]:
lin_pred = lin_reg.predict(X_test)
tree_pred = tree_reg.predict(X_test)

In [40]:
lin_test_rmse = np.sqrt(mean_squared_error(y_test, lin_pred))
tree_test_rmse = np.sqrt(mean_squared_error(y_test, tree_pred))

lin_test_rmse
tree_test_rmse

0.5345224838248488

**4) Fine tune your Decision Tree model by using grid search and then try Random Forest model and do a randomized search.** (4 points in total, see below)
- Report the best hyperparameters and RMSE scores
- 1 point: Decision Tree + Grid search + reporting as asked above
- 1 point: Decision Tree + Randomized search + reporting as asked above
- 1 point: Random Forest + Grid search + reporting as asked above
- 1 point: Random Forest + Randomized search + reporting as asked above

In [41]:
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from scipy.stats import randint

In [43]:
def neg_mse_to_rmse(neg_mse):
    return np.sqrt(-neg_mse)

In [44]:
tree_reg = DecisionTreeRegressor(random_state = 42)

tree_param_grid = {
    "max_depth": [None, 2, 3, 4, 5, 6, 8, 10],
    "min_samples_split": [2, 5, 10, 20],
    "min_samples_leaf": [1, 2, 4, 8],
    "max_features": [None, "sqrt", "log2"]
}

tree_grid = GridSearchCV(
    estimator = tree_reg,
    param_grid = tree_param_grid,
    scoring = "neg_mean_squared_error",
    cv = 5,
    n_jobs = -1
)

tree_grid.fit(X_train, y_train)

tree_grid_best_params = tree_grid.best_params_
tree_grid_best_rmse = neg_mse_to_rmse(tree_grid.best_score_)

tree_grid_best_params, tree_grid_best_rmse

C:\Users\eddie\anaconda3\Lib\site-packages\numpy\ma\core.py:2820: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,


({'max_depth': 4,
  'max_features': None,
  'min_samples_leaf': 8,
  'min_samples_split': 2},
 0.4070689129579489)

In [45]:
tree_rand_dist = {
    "max_depth": randint(1, 21),             
    "min_samples_split": randint(2, 31),     
    "min_samples_leaf": randint(1, 21),      
    "max_features": [None, "sqrt", "log2"]
}

tree_random = RandomizedSearchCV(
    estimator = tree_reg,
    param_distributions = tree_rand_dist,
    n_iter = 50,
    scoring = "neg_mean_squared_error",
    cv = 5,
    random_state = 42,
    n_jobs = -1
)

tree_random.fit(X_train, y_train)

tree_random_best_params = tree_random.best_params_
tree_random_best_rmse = neg_mse_to_rmse(tree_random.best_score_)

tree_random_best_params, tree_random_best_rmse

({'max_depth': 16,
  'max_features': None,
  'min_samples_leaf': 18,
  'min_samples_split': 16},
 0.4045897098179576)

In [46]:
rf_reg = RandomForestRegressor(random_state=42)

rf_param_grid = {
    "n_estimators": [100, 200, 500],
    "max_depth": [None, 5, 10, 20],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4],
    "max_features": ["sqrt", "log2", None],
    "bootstrap": [True, False]
}

rf_grid = GridSearchCV(
    estimator = rf_reg,
    param_grid = rf_param_grid,
    scoring = "neg_mean_squared_error",
    cv = 5,
    n_jobs = -1
)

rf_grid.fit(X_train, y_train)

rf_grid_best_params = rf_grid.best_params_
rf_grid_best_rmse = neg_mse_to_rmse(rf_grid.best_score_)

rf_grid_best_params, rf_grid_best_rmse

({'bootstrap': True,
  'max_depth': 5,
  'max_features': 'log2',
  'min_samples_leaf': 1,
  'min_samples_split': 5,
  'n_estimators': 200},
 0.3904437793973753)

In [47]:
rf_rand_dist = {
    "n_estimators": randint(100, 800),
    "max_depth": [None] + list(range(3, 31)),
    "min_samples_split": randint(2, 31),
    "min_samples_leaf": randint(1, 21),
    "max_features": ["sqrt", "log2", None],
    "bootstrap": [True, False]
}

rf_random = RandomizedSearchCV(
    estimator = rf_reg,
    param_distributions = rf_rand_dist,
    n_iter = 60,
    scoring = "neg_mean_squared_error",
    cv = 5,
    random_state = 42,
    n_jobs = -1
)

rf_random.fit(X_train, y_train)

rf_random_best_params = rf_random.best_params_
rf_random_best_rmse = neg_mse_to_rmse(rf_random.best_score_)

rf_random_best_params, rf_random_best_rmse

({'bootstrap': True,
  'max_depth': 11,
  'max_features': None,
  'min_samples_leaf': 15,
  'min_samples_split': 20,
  'n_estimators': 666},
 0.38930494805031185)

**5) Final Model Evaluation**
- Select the best-performing model based on RMSE. (1 point)
- Evaluate and report **RMSE on the test set** for the best model. (1 point)

In [51]:
results = {
    "Tree Grid": np.sqrt(-tree_grid.best_score_),
    "Tree Random": np.sqrt(-tree_random.best_score_),
    "RF Grid": np.sqrt(-rf_grid.best_score_),
    "RF Random": np.sqrt(-rf_random.best_score_)
}

results

{'Tree Grid': 0.4070689129579489,
 'Tree Random': 0.4045897098179576,
 'RF Grid': 0.3904437793973753,
 'RF Random': 0.38930494805031185}

In [52]:
best_model = rf_random.best_estimator_
best_model

RandomForestRegressor(max_depth=11, max_features=None, min_samples_leaf=15,
                      min_samples_split=20, n_estimators=666, random_state=42)

In [53]:
best_pred = best_model.predict(X_test)

best_test_rmse = np.sqrt(mean_squared_error(y_test, best_pred))

best_test_rmse

0.3997729963445861

**6) What do your results mean?** (1 point)
- What do the results mean in context of the Diabetes data?
- How good the predictions are in your opinion?

Double-click here to write down your answer:
- We set 0 for no Diabetes and 1 for yes Diabetes, but we didnt use classification but regression so the model would show values from 0 to 1.
- RMSE means thw average distance between the predicted value and true value. So in our case the predictions are off by 0.39 on a scale from 0 to 1.
- Our dataset is small, we only used 8 features which is not many, it is a hard prediction so our result of 0.39 is acceptable.
- This results should of course not be relied upon to producre real world usable predictions.